upt_macports/tests/test_macports_package.py

```diff
@@ -32,5 +32,21 @@ def test_multiple_license(self):
         self.assertEqual(self.package.licenses, expected)
 
 
+class TestMacPortsPackageArchiveType(unittest.TestCase):
+    def setUp(self):
+        self.package = MacPortsPackage()
+        self.package.upt_pkg = upt.Package('foo', '42')
+
```

Comment:
Add a test with:

>  self.package.upt_pkg.archives = []

See what happens.

Reply:
It raises an ArchiveUnavailable error as expected, I have added a test for the same.

```
def test_no_archive(self):
        self.package.upt_pkg.archives = []
        with self.assertRaises(upt.ArchiveUnavailable):
            self.package.archive_type
```

Reply:
But should "self.package.archive_type" raise an exception? Shouldn't it rather handle it and return 'unknown'?

Reply:
If we handle the exception and generate the portfile then many details will be missing as checksums will be missing too.
So should we really generate the portfile when the main thing is missing?
I agree that at least a few details will be there and that can be used as an outline by the maintainer.

Reply:
The checksums would be missing, not wrong.

We should probably "do our best". The goal of upt is to create roughly 90% of the package. If for some reason we can only create 50%, well, the maintainer still has something to work with, and saved a few minutes of their life.

We should probably log this as an error, though.

Reply:
> We should probably log this as an error, though.
Yes, that would be good indeed (there are probably other instances to log as error/warning; for example when the license cannot be translated to MacPorts). Should that be part of this PR or should we look at the logging by ```upt_macports``` separately?

Reply:
Ok so if the archive is not found then at this point should we log 'Archive not found on upstream' or should we say 'Archive detection failed as no archive found' as the latter is appropriate as per the place where it is being logged but even the former is not wrong.

And also as of now, I am logging it like this 
```
try:
            archive_url = self.upt_pkg.get_archive().url
            archive_type = archive_url.split('.')[-1]
            return archive_keyword.get(archive_type, 'unknown')
except upt.ArchiveUnavailable:
            self.logger.error(f'Archive detection failed as no archive found')
            return 'unknown'
```
and the test has been changed accordingly.
```
self.package.upt_pkg.archives = []
expected = 'unknown'
self.assertEqual(self.package.archive_type, expected)
```

Is it fine now? :)

Also as of the other logging goes I guess from now on for every feature wherever required we will add the logging and for the existing code, we can add the logging in another PR.

Reply:
hi @Korusuke, ehm I don't really see the change you mention above in the PR... but perhaps you were waiting for a response before adding it?

I would like to avoid two ```return``` statements, so what about:
```
try:
    archive_url = self.upt_pkg.get_archive().url
    archive_type = archive_url.split('.')[-1]
except upt.ArchiveUnavailable:
    self.logger.error(f'Detection of archive type failed: no archive found')
    archive_type = None
finally:
    return archive_keyword.get(archive_type, 'unknown')
```

Additionally, as I commented earlier I would not treat ```gz``` differently in the ```archive_keyword``` dictionary than ```xz```; just return ```gz``` there as well.

Reply:
So we could use "self.upt_pkg.get_archive().filename" rather than "self.upt_pkg.get_archive().url", but both work :)

I like the simple try/except better than the try/except/finally with a single return statement, but I'm ok with both as long as we have good tests.

Reply:
fine with me, leave the ```try....except``` as is, but please do change the logging message. Using ```self.upt_pkg.get_archive().filename``` would indeed make sense.

Reply:
Yeah using `self.upt_pkg.get_archive().filename` make more sense, I had missed that it exists as my local copy of upt was old and did not have the filename, filepath properties :)